### PR TITLE
signup

### PR DESCRIPTION
**Pull request by MinhDo2007**

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class UsersController < ApplicationController
+  def new
+  	@user = User.new
```
Comment (khanhkhanh) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      redirect_to root_url
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit :name, :email, :password,
+                                   :password_confirmation
```
Comment (khanhkhanh) [inline app/controllers/users_controller.rb]: :name and :password_confirmation have the same indentation

File: app/helpers/users_helper.rb [hunk 3]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (khanhkhanh) [inline app/helpers/users_helper.rb]: Remove this file if not use

File: app/models/user.rb [hunk 4]
```diff
@@ -8,4 +8,14 @@ class User < ActiveRecord::Base
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
+  
+  #check validate
+  
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :email, presence: true, length: { maximum: 255 },
```
Comment (khanhkhanh) [inline app/models/user.rb]: Remove spaces after { and before }

File: app/views/shared/_error_messages.html.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (khanhkhanh) [inline app/views/shared/_error_messages.html.erb]: i18n

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<% provide :title, t("signup") %>
+<h1>Sign up</h1>
```
Comment (khanhkhanh) [inline app/views/users/new.html.erb]: i18n

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (khanhkhanh) [inline app/views/users/new.html.erb]: Remove ()

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: "form-control" %>
+
+      <%= f.submit "Create my account", class: "btn btn-primary" %>
```
Comment (khanhkhanh) [inline app/views/users/new.html.erb]: i18n

File: app/views/users/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    <% end %>
+  </div>
+</div>
```
Comment (khanhkhanh) [inline app/views/users/new.html.erb]: Break line

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (phamkykhoi) [inline app/controllers/users_controller.rb]: Indent

File: app/views/users/new.html.erb (hunk 6 above)
Comment (MinhDo2007) [inline app/views/users/new.html.erb]: t("signup") i18n

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
```
Comment (nguyenduyta) [inline app/controllers/users_controller.rb]: cách 1 dòng ra anh ạ

File: app/models/user.rb [hunk 11]
```diff
@@ -8,4 +8,14 @@ class User < ActiveRecord::Base
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
+  
+  #check validate
```
Comment (nguyenduyta) [inline app/models/user.rb]: remove comment

File: app/views/shared/_error_messages.html.erb [hunk 12]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, t("error")) %>.
```
Comment (nguyenduyta) [inline app/views/shared/_error_messages.html.erb]: I18n

File: app/models/user.rb [hunk 13]
```diff
@@ -8,4 +8,14 @@ class User < ActiveRecord::Base
...
+  
+  before_save {self.email = email.downcase}
```
Comment (nguyenduyta) [inline app/models/user.rb]: với mỗi block anh nên phân cách bởi 1 dòng trắng.
các quan hệ hasmany, hasone, belongsto thành 1 block. validate 1 block. function after_save, before_save... 1 block. khi đọc code sẽ dễ hơn ạ

File: app/views/shared/_error_messages.html.erb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
```
Comment (phamkykhoi) [inline app/views/shared/_error_messages.html.erb]: Should use object get from form user view

File: app/views/shared/_error_messages.html.erb [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t("contains") %><%= pluralize(@user.errors.count, t("error")) %>.
+    </div>
+    <ul>
+    <% @user.errors.full_messages.each do |msg| %>
```
Comment (phamkykhoi) [inline app/views/shared/_error_messages.html.erb]: Should use object get from form user view

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (nguyenduyta) [inline app/controllers/users_controller.rb]: Add new line blank

File: app/models/user.rb [hunk 16]
```diff
@@ -8,4 +8,13 @@ class User < ActiveRecord::Base
                                   dependent: :destroy
   has_many :following, through: :active_relationships, source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
+  
+  before_save {self.email = email.downcase}
+
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+  validates :name, presence: true, length: {maximum: 50}
```
Comment (phamkykhoi) [inline app/models/user.rb]: maximum: 50  Move Settings

File: app/models/user.rb [hunk 17]
```diff
@@ -8,4 +8,13 @@ class User < ActiveRecord::Base
...
+  validates :email, presence: true, length: {maximum: 255},
```
Comment (phamkykhoi) [inline app/models/user.rb]: maximum: 250  Move Settings

File: app/views/shared/_error_messages.html.erb [hunk 18]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t("contains") %><%= pluralize(object.errors.count, t("error")) %>.
```
Comment (khanhkhanh) [inline app/views/shared/_error_messages.html.erb]: <%= t("contains") %><%= pluralize(object.errors.count, t("error")) %>
=> <%= t "contains" %><%= pluralize object.errors.count, t("error") %>

File: app/views/static_pages/home.html.erb [hunk 19]
```diff
@@ -1,6 +1,6 @@
-<h1><%= t("home") %></h1>
+<h1><%= t "home" %></h1>
 <div class="center jumbotron">
   <h1><%= t "welcome" %></h1>
   <h2><%= t "home-description" %></h2>
-  <%= link_to t("signup-button"), "#", class: "btn btn-lg btn-primary" %>
+  <%= link_to t("signup-button"), "users/new", class: "btn btn-lg btn-primary" %>
```
Comment (khanhkhanh) [inline app/views/static_pages/home.html.erb]: "users/new" => Use path instead

File: app/views/shared/_error_messages.html.erb [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+    </div>
+    <ul>
+    <% object.errors.full_messages.each do |msg| %>
+      <li><%= msg %></li>
+    <% end %>
+    </ul>
+  </div>
+<% end %>
```
Comment (khanhkhanh) [inline app/views/shared/_error_messages.html.erb]: Break line

File: app/views/static_pages/help.html.erb [hunk 21]
```diff
@@ -1,5 +1,5 @@
 <% provide :title, t(:help) %>
-<h1><%= t :help %></h1>
+<h1><%= t "help" %></h1>
 <div class="center jumbotron">
   <h2><%= t :"help-description" %></h2>
```
Comment (khanhkhanh) [inline app/views/static_pages/help.html.erb]: t :"help-description" => t :help-description

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (MinhDo2007) [inline app/controllers/users_controller.rb]: I have a new line at line 5

File: app/views/shared/_error_messages.html.erb (hunk 20 above)
Comment (MinhDo2007) [inline app/views/shared/_error_messages.html.erb]: where break line?

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (nguyenduyta) [inline app/controllers/users_controller.rb]: okie

**Comment by nguyenduyta** [comment]
anh xóa 3 file 
config/settings/development.yml
config/settings/production.yml
config/settings/test.yml
đi nhé

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (MinhDo2007) [inline app/controllers/users_controller.rb]: pls , delete your cmt @@

File: app/models/user.rb [hunk 22]
```diff
@@ -8,4 +8,14 @@ class User < ActiveRecord::Base
...
   has_many :following, through: :active_relationships, source: :followed
   has_many :followers, through: :passive_relationships, source: :follower
+  
+  before_save {self.email = email.downcase}
+
+  validates :name, presence: true, length: {maximum: Settings.setting.maximum_length_name}
+  validates :email, presence: true, length: {maximum: Settings.setting.maximum_length_email},
+                    format: {with: Settings.setting.VALID_EMAIL_REGEX},
+                    uniqueness: {case_sensitive: false}
+  validates :password, presence: true, length: { minimum: Settings.setting.minimum_length_password }
```
Comment (phamkykhoi) [inline app/models/user.rb]: Remove space {  }

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+  
+  def create
+    @user = User.new(user_params)
```
Comment (dieunb) [inline app/controllers/users_controller.rb]: remove `()`

File: app/views/users/new.html.erb [hunk 24]
```diff
@@ -0,0 +1,22 @@
+<% provide :title, t("signup") %>
+<h1><%= t "signup" %></h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
```
Comment (dieunb) [inline app/views/users/new.html.erb]: slice to new partial file

File: Gemfile [hunk 25]
```diff
@@ -10,6 +10,8 @@ gem "jquery-rails"
 gem "turbolinks"
 gem "jbuilder", "~> 2.0"
 gem "sdoc", "~> 0.4.0", group: :doc
+gem 'bcrypt', "~> 3.1.7"
```
Comment (dieunb) [inline Gemfile]: `"`, not `'`

**Comment by dieunb** [comment]
OK

**Pull request merged**